src/QsCompiler/SyntaxProcessor/ExpressionVerification.fs

```diff
@@ -396,19 +396,23 @@ let private characteristicsSet info =
         | Controlled -> Controllable)
     |> Set.ofSeq
 
-let private lambdaCharacteristics (body: TypedExpression) =
-    let mutable characteristics = Set.ofList [ Adjointable; Controllable ]
-
-    let onCall =
-        function
+let private lambdaCharacteristics (inference: InferenceContext) (body: TypedExpression) =
+    let mutable characteristics =
+        if inference.Resolve(body.ResolvedType).Resolution = UnitType then
+            Set.ofList [ Adjointable; Controllable ]
+        else
+            Set.empty
+
+    let onCall callableType =
+        match inference.Resolve(callableType).Resolution with
         | QsTypeKind.Operation (_, info) -> characteristics <- characteristicsSet info |> Set.intersect characteristics
         | TypeParameter _ -> characteristics <- Set.empty
```

Comment:
I've not been able to figure this match case out yet. Why do we need special handling for type parameters?

Reply:
I added a comment about this:

```fsharp
// When a callable type can't be resolved to an operation type based on the current knowledge of the
// inference context, pessimistically assume that it is an operation that supports no characteristics. This
// limitation exists by design to make characteristics inference easier.
```

Does it make sense?

Reply:
Okay so the type parameter case is if the callable is of some type parameter 'T that is a callable type? In that case, we don't have enough information about the 'T to know what characteristics it supports, so we assume it doesn't support any to be safe. Is that right? Is there some syntax in Q# that let's us know that 'T must be a callable type, and if so is there an extension of that syntax that can let us enforce that the type has certain characteristics?

Reply:
A declared type parameter can never be called, so that is actually not the case we need to prevent. The problematic case is something like this:

```qsharp
use q = Qubit();
let lambda = op => op(q);
lambda(X);
```

At the point where the type of `op => op(q)` is being inferred, `op` is some auto-generated type variable, call it `'a`. We have to choose some characteristics for `lambda` but we don't know what `'a` is without looking ahead to the next line, `lambda(X)`. For simplicity we just assume `'a` is an operation with no characteristics, even though that is not the only way to type this program.